### DISCUSSION TITLE
How to identify the different outputs

### DISCUSSION THREAD

**Discussion by tv4you2016** [HomeSpan/HomeSpan#963] [Q&A]
Hello.
I'm creating a hub, and follow the example:
ProgrammableHub.ino and added the [Dev_LED.txt](https://github.com/user-attachments/files/18126482/Dev_LED.txt)  file 
I only edited the following lines: 

changes: 
```
    new Service::LightBulb();
      new Characteristic::On(0,true);
      if(lightData[index].isDimmable)
        new Characteristic::Brightness(100,true);
      if(lightData[index].colorType==TEMPERATURE_ONLY)
        new Characteristic::ColorTemperature(200,true);        
      if(lightData[index].colorType==FULL_RGB){
        new Characteristic::Hue(0,true);
        new Characteristic::Saturation(0,true);
```
    
to: 

` new DEV_LED(index,lightData[index].isDimmable,lightData[index].colorType);`

in 

```
 boolean update(){ // update() method
 
    log_i("ON: %s",characteristic->getNewVal()?"true":"false");

    return(true);                               // return true
  
  } // update

```
How can I identify which device you changed??

**Comment by tv4you2016** [comment]
@HomeSpan 
any ideas that could help me?

**Comment by HomeSpan** [comment]
You may be jumping into a complex example a little too fast.  Have you gone through the basic examples, including those that show multiple LEDs.  As you'll see, the update command is tied to each individual LED object, so whatever info is available in each instance of the object is available to the update command, including whatever variables you've added to specify pin numbers or other identifiers.

**Comment by tv4you2016** [comment]
Hi @HomeSpan , you were right, I started running all the examples and when I get to the 16-ProgrammableSwitches example, I encountered a difficulty, I put the example just like Button PushButton Switches is detected in Homekit, but when I click on the button pressType LONG, Simgle or double is not detected in homekit despite the event being triggered


`ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x1f (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13232
load:0x40080400,len:3028
entry 0x400805e4
[    42][D][esp32-hal-cpu.c:244] setCpuFrequencyMhz(): PLL: 480 / 2 = 240 Mhz, APB: 80000000 Hz
[   500][I][esp32-hal-psram.c:96] psramInit(): PSRAM enabled
[   507][V][esp32-hal-uart.c:330] uartBegin(): UART0 baud(115200) Mode(800001c) rxPin(3) txPin(1)
[   519][V][esp32-hal-uart.c:416] uartBegin(): UART0 not installed. Starting installation
[   530][V][esp32-hal-uart.c:467] uartBegin(): UART0 initialization done.
[   653][V][WiFiServer.h:42] WiFiServer(): WiFiServer::WiFiServer(port=80, ...)

************************************************************
Welcome to HomeSpan!
Apple HomeKit for the Espressif ESP-32 WROOM and Arduino IDE
************************************************************

** Please ensure serial monitor is set to transmit <newlines>

Message Logs:     Level 2
Status LED:       Pin -  *** WARNING: Status LED Pin is UNDEFINED
Device Control:   Pin -  *** WARNING: Device Control Pin is UNDEFINED
Sketch Version:   n/a
HomeSpan Version: 1.9.1
Arduino-ESP Ver.: 2.0.17
ESP-IDF Version:  4.4.7
ESP32 Chip:       ESP32-D0WD Rev 1 dual-core 4MB Flash
ESP32 Board:      esp32
PWM Resources:    16 channels, 8 timers, max 20-bit duty resolution
Sodium Version:   1.0.12-idf  Lib 9.4
MbedTLS Version:  Mbed TLS 2.28.7
Sketch Compiled:  Dec 15 2024 09:55:36
Partition:        app0
MAC Address:      40:F5:20:79:5D:F4

Device Name:      HomeSpan Bridge

Configuring Programmable Pushbutton: Pin=23  Index=1
Configuring Programmable Pushbutton: Pin=5  Index=2

*** HomeSpan Info ***

➤ Accessory:  AID=1
   ➟ Service AccessoryInformation:  IID=1, UUID="3E"
      ⇨ Characteristic Identify(1):  IID=2, UUID="14", Perms=PW
➤ Accessory:  AID=2
   ➟ Service AccessoryInformation:  IID=1, UUID="3E"
      ⇨ Characteristic Identify(1):  IID=2, UUID="14", Perms=PW
      ⇨ Characteristic Name("PushButton Switches"):  IID=3, UUID="23", Perms=PR
   ➟ Service StatelessProgrammableSwitch:  IID=4, UUID="89"
      ⇨ Characteristic ProgrammableSwitchEvent(0):  IID=5, UUID="73", Perms=PR+EV, Range=[0,2], EV=()
      ⇨ Characteristic ServiceLabelIndex(1):  IID=6, UUID="CB", Perms=PR, Range=[1,255]
      ▼ SpanButton: Pin=23, Single=5ms, Double=200ms, Long=2000ms, Type=TRIGGER_ON_LOW
   ➟ Service StatelessProgrammableSwitch:  IID=7, UUID="89"
      ⇨ Characteristic ProgrammableSwitchEvent(0):  IID=8, UUID="73", Perms=PR+EV, Range=[0,2], EV=()
      ⇨ Characteristic ServiceLabelIndex(2):  IID=9, UUID="CB", Perms=PR, Range=[1,255]
      ▼ SpanButton: Pin=5, Single=5ms, Double=200ms, Long=2000ms, Type=TRIGGER_ON_LOW

Service                             UUID         AID  IID  Update  Loop  Button  Linked Services
------------------------------  --------  ----------  ---  ------  ----  ------  ---------------
AccessoryInformation                  3E           1    1      NO    NO      NO  -
AccessoryInformation                  3E           2    1      NO    NO      NO  -
StatelessProgrammableSwitch           89           2    4      NO    NO     YES  -
StatelessProgrammableSwitch           89           2    7      NO    NO     YES  -

Configured as Bridge: YES

Database Validation:  Warnings=0, Errors=0

*** End Info ***

Accessory ID:      80:45:8D:ED:A8:9F                               LTPK: 4F3B2F9E9C183EE03B9E4981C3CAC6183B5DD4495D0138298483A574F1E6D158
Paired Controller: 9E69A841-A5B3-4B98-AD5B-1B45118B65EB   (admin)  LTPK: F093D4BB4E32AEBB4D013745F1951752F1544A8D4AEECE360908F6464E706602

Accessory configuration number: 13

HomeSpan Bridge is READY!

Trying to connect to Isso_Agora.  Waiting 1 sec...
[  2942][D][WiFiGeneric.cpp:1040] _eventCallback(): Arduino Event: 0 - WIFI_READY
[  3023][V][WiFiGeneric.cpp:97] [e0t;_3e7smp[_ i n3t0e2r1f]a[cVe]_[iWpi(F)i:G eCnoenrfiicg.ucrpipn:g3 4S1t]a t_iaornd ustno_c ent c.0.: S,A SSK:ted0.0m
 GW: 0.0.0.0
[  3039][D][WiFiGeneric.cpp:1040] _eventCallback(): Arduino Event: 2 - STA_START
[  3093][V][WiFiGeneric.cpp:356] _arduino_event_cb(): STA Connected: SSID: Isso_Agora, BSSID: 04:42:1a:7b:7d:b8, Channel: 7, Auth: WPA2_PSK
[  3107][D][WiFiGeneric.cpp:1040] _eventCallback(): Arduino Event: 4 - STA_CONNECTED
[  3121][V][WiFiGeneric.cpp:370] _arduino_event_cb(): STA Got New IP:192.168.1.244
[  3130][D][WiFiGeneric.cpp:1040] _eventCallback(): Arduino Event: 7 - STA_GOT_IP
[  3138][D][WiFiGeneric.cpp:1103] _eventCallback(): STA IP: 192.168.1.244, MASK: 255.255.255.0, GW: 192.168.1.2
WiFi Connected!  IP Address = 192.168.1.244

Starting MDNS...

HostName:      HomeSpan-80458DEDA89F.local:80
Display Name:  HomeSpan Bridge
Model Name:    HomeSpan-ESP32
Setup ID:      HSPN

Starting HAP Server on port 80...


Found button press on pin: 5  type: SINGLE
Found button press on pin: 5  type: SINGLE
Found button press on pin: 5  type: SINGLE
Found button press on pin: 5  type: SINGLE
=======================================
** Client #1 Connected (15 sec): 192.168.1.218

<<<<<<<<< 192.168.1.218 <<<<<<<<<
POST /pair-verify HTTP/1.1
Host: HomeSpan\032Bridge._hap._tcp.local
Content-Length: 37
Content-Type: application/pairing+tlv8
------------ END BODY! ------------
STATE(1) 01 [1]
PUBKEY(32) 824514FC7AB0D4FC834D3E85516273E459C261349F12F8326BB4780E7F91816B
------------ END TLVS! ------------
In Pair Verify #1 (192.168.1.218)...Found <M1>
------- ENCRYPTING SUB-TLVS -------
IDENTIFIER(17) 38303A34353A38443A45443A41383A3946
SIGNATURE(64) FAD1CFA1432DAAC3E3D88CF0D956F55BE4035C22E15424ACE41EF5C87095736B86AA8002553AAF6EAA41E8BE89C22BDB41E9A8BB51F8889FB44799BF658AEB06
---------- END SUB-TLVS! ----------

>>>>>>>>>> 192.168.1.218 >>>>>>>>>>
HTTP/1.1 200 OK
Content-Type: application/pairing+tlv8
Content-Length: 140

ENC.DATA(101) FD8296FD7A4FF26FB7A4F6BCEEA30A63B45559666B0C0223C4ED8C6DCB341B4EBAEFC9208148F316DFEA8863E571B3DE9E53E935E2E6EE5B0F30A33E7805932A9EAA03AB0F3913C95590AE567548F087662DA27FA0A30A01587212849C1D012F08398B6A35
STATE(1) 02 [2]
PUBKEY(32) DA5ADF73E0EB7A7A3A80BEF9AEDF4545127CCFEF03507FC8D2F93AD4D78DDF60
------------ SENT! --------------
<<<<<<<<< 192.168.1.218 <<<<<<<<<
POST /pair-verify HTTP/1.1
Host: HomeSpan\032Bridge._hap._tcp.local
Content-Length: 125
Content-Type: application/pairing+tlv8
------------ END BODY! ------------
ENC.DATA(120) 42C10B50DF452BD98B2118C6634210EC6423C90CB54BBCB60E8DD2E07588EA83F0210228F97D8A510700619D6EB47E0B4F29AD9C9FAEEDB1F401A0DCE529AF71B5867D311933201A76DD6502E350D2CE9866A1F1D1349A586FE20DA5A4ED27FCB4833148D3E63677F7B4C77AECB5C37F0075665C45ECE70E
STATE(1) 03 [3]
------------ END TLVS! ------------
In Pair Verify #1 (192.168.1.218)...Found <M3>
------- DECRYPTING SUB-TLVS -------
IDENTIFIER(36) 39453639413834312D413542332D344239382D414435422D314234353131384236354542
SIGNATURE(64) A62D8B7FB8216617D3F7822414C51EB61335515552F383AEC0D612F2379494938ACCDD5E62C97D8D43E293A4564C10F84BF7FD51FBCC48DABA3F4A197E96B309
---------- END SUB-TLVS! ----------

*** Verifying session with Controller ID: 9E69A841-A5B3-4B98-AD5B-1B45118B65EB...

>>>>>>>>>> 192.168.1.218 >>>>>>>>>>
HTTP/1.1 200 OK
Content-Type: application/pairing+tlv8
Content-Length: 3

STATE(1) 04 [4]
------------ SENT! --------------

*** SESSION VERIFICATION COMPLETE ***
<<<< #### 192.168.1.218 #### <<<<
GET /accessories HTTP/1.1
Host: HomeSpan\032Bridge._hap._tcp.local
------------ END BODY! ------------
In Get Accessories #1 (192.168.1.218)...

>>>>>>>>>> 192.168.1.218 >>>>>>>>>>
HTTP/1.1 200 OK
Content-Type: application/hap+json
Content-Length: 696

{"accessories":[{"aid":1,"services":[{"iid":1,"type":"3E","characteristics":[{"iid":2,"type":"14","format":"bool","perms":["pw"]}]}]},{"aid":2,"services":[{"iid":1,"type":"3E","characteristics":[{"iid":2,"type":"14","format":"bool","perms":["pw"]},{"iid":3,"type":"23","value":"PushButton Switches","format":"string","perms":["pr"]}]},{"iid":4,"type":"89","characteristics":[{"iid":5,"type":"73","value":null,"format":"uint8","perms":["pr","ev",]},{"iid":6,"type":"CB","value":1,"format":"uint8","perms":["pr"]}]},{"iid":7,"type":"89","characteristics":[{"iid":8,"type":"73","value":null,"format":"uint8","perms":["pr","ev",]},{"iid":9,"type":"CB","value":2,"format":"uint8","perms":["pr"]}]}]}]}
-------- SENT ENCRYPTED! --------
<<<< #### 192.168.1.218 #### <<<<
PUT /characteristics HTTP/1.1
Host: HomeSpan\032Bridge._hap._tcp.local
Content-Length: 77
Content-Type: application/hap+json
------------ END BODY! ------------
{"characteristics":[{"aid":2,"iid":5,"ev":true},{"aid":2,"iid":8,"ev":true}]}
------------ END JSON! ------------
In Put Characteristics #1 (192.168.1.218)...
Notification Request for aid=2 iid=5: true
Notification Request for aid=2 iid=8: true

>>>>>>>>>> 192.168.1.218 >>>>>>>>>>
HTTP/1.1 204 No Content


-------- SENT ENCRYPTED! --------
Found button press on pin: 5  type: SINGLE

>>>>>>>>>> 192.168.1.218 >>>>>>>>>>
EVENT/1.0 200 OK
Content-Type: application/hap+json
Content-Length: 49

{"characteristics":[{"iid":8,"value":0,"aid":2}]}
-------- SENT ENCRYPTED! --------
Found button press on pin: 5  type: SINGLE

>>>>>>>>>> 192.168.1.218 >>>>>>>>>>
EVENT/1.0 200 OK
Content-Type: application/hap+json
Content-Length: 49

{"characteristics":[{"iid":8,"value":0,"aid":2}]}
-------- SENT ENCRYPTED! --------
[ 25688][D][WiFiClient.cpp:536] connected(): Disconnected: RES: 0, ERR: 128
** Client #1 DISCONNECTED (25 sec)
Found button press on pin: 5  type: SINGLE
Found button press on pin: 5  type: SINGLE
Found button press on pin: 5  type: SINGLE
Found button press on pin: 5  type: SINGLE
Found button press on pin: 5  type: SINGLE
Found button press on pin: 5  type: SINGLE
Found button press on pin: 5  type: SINGLE
Found button press on pin: 5  type: DOUBLE
Found button press on pin: 5  type: DOUBLE
Found button press on pin: 5  type: LONG`

**Comment by HomeSpan** [comment]
According to the log above, your button presses are being sent (it's the line that says "EVENT") but after a few times HomeKit is dropping the connection to HomeSpan.  This is beyond the control of HomeSpan.  

Do you have a HomePod or AppleTV configured to run as a HomeKit Hub?  This is a MUST if you want to use button presses.
